### PR TITLE
Fix dynamically declared fields

### PR DESCRIPTION
**Pull request by sankalpshekhar**
### Description (*)

Add Missing fields to the Core classes

### Fixed Issues (if relevant)

1. magento/magento2#<issue_number>: Issue title

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Backend/Model/Menu/Config.php
Comment (dmytro-ch) [inline app/code/Magento/Backend/Model/Menu/Config.php:71]: Please add `_menuBuilder` property with the `private` scope of visibility.

**Review by dmytro-ch — CHANGES_REQUESTED** [review]
Hi @sankalpshekhar,
thank you for your contribution!
Could you please check and resolve the automated tests failures?

**Review by orlangur — CHANGES_REQUESTED** [review]
@sankalpshekhar we don't need such occurrences fixed in one module. Please fix them all at once with the help of automated tooling.

File: app/code/Magento/Backend/Model/Menu/Config.php
Comment (sankalpshekhar) [reply app/code/Magento/Backend/Model/Menu/Config.php:71]: @dmytro-ch I understand as per Magento's guidelines every field needs to be made private, however every-other field on the class seems to been marked protected, hence i marked this too as protected. Let me know if you still believe this needs to be marked as private.

**Comment by orlangur** [comment]
@sankalpshekhar, "Please fix them all at once with the help of **automated** tooling."

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=23102) <br/>All committers have signed the CLA.

**Review by orlangur — CHANGES_REQUESTED** [review]
@sankalpshekhar nice! Please provide exact steps with automated tool of your choice so that they can be reproduced for code review.

**Comment by sankalpshekhar** [comment]
@orlangur  Please see the steps involved using PHPStorm.

1. Create  new inspection profile in Settings > Editors > Inspections
2. Duplicate the default one, disable all except one for PHP > Undefined > Undefined Field
3. Save profile
4. Select Magento directory under app > code in the project window, and execute inspection to get the result

**Comment by sankalpshekhar** [comment]
@orlangur Kindly let know, on what else is remaining to get closure on this.

**Review by sidolov — CHANGES_REQUESTED** [review]
Hi @sankalpshekhar , we cannot declare dynamically fields as private because it's breaking changes. By default dynamically declared filed has public visibility modifier, that means that property accessible from in the all child classes, declaring such properties as private we will eliminate this behavior. I'll suggest you declare mentioned properties at least with protected visibility modifier to avoid breaking changes.

**Comment by sankalpshekhar** [comment]
> 
> 
> Please add `_menuBuilder` property with the `private` scope of visibility.


@sidolov I tried the same, but i was requested to turn this to private.

**Comment by sidolov** [comment]
@sankalpshekhar , yes, we should declare new properties as private, but we can't limit property visibility if it was defined before (in our case it's public by default)

File: app/code/Magento/Backend/Model/Menu/Config.php
Comment (YevSent) [reply app/code/Magento/Backend/Model/Menu/Config.php:71]: @sankalpshekhar all properties by default should be private. The protected properties provide a possibility to extend the class but the composition should be used instead of inheritance. These protected properties are legacy fields.

Comment (sankalpshekhar) [reply app/code/Magento/Backend/Model/Menu/Config.php:71]: @joni-jones I understnad that, but refer to @sidolov comment down below.

> By default dynamically declared filed has public visibility modifier, that means that property accessible from in the all child classes, declaring such properties as private we will eliminate this behavior. I'll suggest you declare mentioned properties at least with protected visibility modifier to avoid breaking changes.

**Comment by sivaschenko** [comment]
@magento run all tests

File: app/code/Magento/Reports/Block/Adminhtml/Grid/AbstractGrid.php
Comment (orlangur) [inline app/code/Magento/Reports/Block/Adminhtml/Grid/AbstractGrid.php]: Why type hint is missing for some of the fields?

**Review by orlangur — CHANGES_REQUESTED** [review]

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @sankalpshekhar,
Thank you for contribution! This PR got stuck, will try to push it forward.

Could you sign Adobe CLA, fix merge conflicts and / or enable edits for maintainers? So we'll be able to fix some issues ourselves
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

File: app/code/Magento/Backend/Model/Menu/Config.php
Comment (ihor-sviziev) [reply app/code/Magento/Backend/Model/Menu/Config.php:71]: Looks reasonable for me to keep it protected. I'm resolving this conversation

File: app/code/Magento/Catalog/Observer/Compare/BindCustomerLoginObserver.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Observer/Compare/BindCustomerLoginObserver.php]: For observers we can do backward incompatible change, so please make this property private.
More info https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

File: app/code/Magento/Catalog/Observer/Compare/BindCustomerLogoutObserver.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Observer/Compare/BindCustomerLogoutObserver.php]: For observers we can do backward incompatible change, so please make this property private.
More info https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

File: app/code/Magento/Catalog/Plugin/Model/Indexer/Category/Product/MaxHeapTableSizeProcessorOnFullReindex.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Plugin/Model/Indexer/Category/Product/MaxHeapTableSizeProcessorOnFullReindex.php]: For plugins we can do backward incompatible change, so please make this property private.
More info https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

File: app/code/Magento/CatalogSearch/Model/Adapter/Mysql/Plugin/Aggregation/Category/DataProvider.php
Comment (ihor-sviziev) [inline app/code/Magento/CatalogSearch/Model/Adapter/Mysql/Plugin/Aggregation/Category/DataProvider.php]: Please make it private, as this is plugin

File: app/code/Magento/Customer/Ui/Component/Listing/Columns.php
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Ui/Component/Listing/Columns.php]: Please revert this change as it removes declared param type

File: app/code/Magento/PageCache/Model/App/CacheIdentifierPlugin.php
Comment (ihor-sviziev) [inline app/code/Magento/PageCache/Model/App/CacheIdentifierPlugin.php]: please make properties with private visibility in plugin

File: app/code/Magento/Sales/Controller/Adminhtml/Order/Creditmemo/AddComment.php
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Controller/Adminhtml/Order/Creditmemo/AddComment.php:56]: By changing alias in this file you broke following import.

File: app/code/Magento/Search/Ui/Component/Listing/Column/Website/Options.php
Comment (ihor-sviziev) [inline app/code/Magento/Search/Ui/Component/Listing/Column/Website/Options.php:41]: Looks like in this class there options variable is only written and never used for read, maybe it's better to remove writing to it?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@sankalpshekhar 
Please review my comments. 
I see that you still didn't sign Adobe CLA, unfortunately without signing it - we can't accept your PR.
Also please review failing tests - looks like your changes broke some functionality, so need to double check if everything is correct.

File: app/code/Magento/Catalog/Observer/Compare/BindCustomerLoginObserver.php
Comment (sankalpshekhar) [reply app/code/Magento/Catalog/Observer/Compare/BindCustomerLoginObserver.php]: Resolved

File: app/code/Magento/Catalog/Observer/Compare/BindCustomerLogoutObserver.php
Comment (sankalpshekhar) [reply app/code/Magento/Catalog/Observer/Compare/BindCustomerLogoutObserver.php]: Resolved

File: app/code/Magento/Catalog/Plugin/Model/Indexer/Category/Product/MaxHeapTableSizeProcessorOnFullReindex.php
Comment (sankalpshekhar) [reply app/code/Magento/Catalog/Plugin/Model/Indexer/Category/Product/MaxHeapTableSizeProcessorOnFullReindex.php]: Resolved

File: app/code/Magento/CatalogSearch/Model/Adapter/Mysql/Plugin/Aggregation/Category/DataProvider.php
Comment (sankalpshekhar) [reply app/code/Magento/CatalogSearch/Model/Adapter/Mysql/Plugin/Aggregation/Category/DataProvider.php]: Resolved

File: app/code/Magento/Customer/Ui/Component/Listing/Columns.php
Comment (sankalpshekhar) [reply app/code/Magento/Customer/Ui/Component/Listing/Columns.php]: Resolved

File: app/code/Magento/PageCache/Model/App/CacheIdentifierPlugin.php
Comment (sankalpshekhar) [reply app/code/Magento/PageCache/Model/App/CacheIdentifierPlugin.php]: Resolved

File: app/code/Magento/Sales/Controller/Adminhtml/Order/Creditmemo/AddComment.php
Comment (sankalpshekhar) [reply app/code/Magento/Sales/Controller/Adminhtml/Order/Creditmemo/AddComment.php:56]: Resolved

File: app/code/Magento/Search/Ui/Component/Listing/Column/Website/Options.php
Comment (sankalpshekhar) [reply app/code/Magento/Search/Ui/Component/Listing/Column/Website/Options.php:41]: Understood, but for backward compatibility this might be needed in case someone has a preference on this and is expecting the value in the mentioned field, which not being defined is public in access.

**Comment by sankalpshekhar** [comment]
@ihor-sviziev I am working with my Employer's legal team before signing the CLA.

I understand my changes won't be merged, but i can continue working based on your feedback and various tests results here, while I wait for a feedback.

**Comment by sankalpshekhar** [comment]
@akaplya As part of static tests, PHPStan Suggests the Magento\Catalog\Model\Config::_init() is undefined.

On checking it seems the method _construct() doesn't applies to this Model any more.

Would it be fine, if i remove this method?

**Comment by sankalpshekhar** [comment]
@sidolov @sivaschenko @ihor-sviziev Automates tests suggests failure on Functional tests, however there is no build report to see whats failing. How can we determine the reason of failure? or if the tests are correct?

Similarly Unit tests shows up as failure, but the report within shows 99.5% pas and others skipped, and doesn't provide support on identifying whats failed here.

File: app/code/Magento/Wishlist/Model/Config.php
Comment (danmooney2) [inline app/code/Magento/Wishlist/Model/Config.php:44]: Please use `private $sharingTextLimit;`.  Avoid protected members at all costs

Comment (ihor-sviziev) [reply app/code/Magento/Wishlist/Model/Config.php:44]: That would be backward incompatible change, so no

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Wishlist/Model/Config.php
Comment (danmooney2) [reply app/code/Magento/Wishlist/Model/Config.php:44]: Ah I see it's dynamically declared field.  Should be public then?

Comment (ihor-sviziev) [reply app/code/Magento/Wishlist/Model/Config.php:44]: In general it should be private, but if we’ll do that - this variable will not be available in classes that extend this one. In order to make it available we decided to make them protected, so this variable will not be available outside of the class

**Comment by slavvka** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Hi @sankalpshekhar,
Will you be able to sign Adobe CLA and finalize your PR?

**Comment by ihor-sviziev** [comment]
@sankalpshekhar Unfortunately I need to close this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**
